mapper/mapper.go

```diff
@@ -638,6 +665,72 @@ func mapCorrectionAlert(ver *dataset.Version, version *sharedModel.Version) {
 	}
 }
 
+func mapCensusOptionsToDimensions(dims []dataset.VersionDimension, opts []dataset.Options, queryStrValues []string, path string) []sharedModel.Dimension {
+	dimensions := []sharedModel.Dimension{}
+	for _, opt := range opts {
+		var pDim sharedModel.Dimension
+
+		for _, dimension := range dims {
+			if dimension.Name == opt.Items[0].DimensionID {
+				pDim.Name = dimension.Name
+				pDim.Description = dimension.Description
+				pDim.IsAreaType = helpers.IsBoolPtr(dimension.IsAreaType)
+				pDim.ShowChange = pDim.IsAreaType
+				pDim.Title = dimension.Label
+				pDim.ID = dimension.ID
+			}
+		}
+
+		pDim.TotalItems = opt.TotalCount
+		midFloor, midCeiling := getTruncationMidRange(opt.TotalCount)
+
+		for i, val := range opt.Items {
+			if pDim.TotalItems > 9 && !helpers.HasStringInSlice(pDim.ID, queryStrValues) {
+				if isTruncationIndex(i, midFloor, midCeiling, len(opt.Items)) {
+					pDim.Values = append(pDim.Values, val.Label)
+				}
+				pDim.IsTruncated = true
+			} else {
+				pDim.Values = append(pDim.Values, val.Label)
+				pDim.IsTruncated = false
+			}
+		}
+
+		q := url.Values{}
+		if pDim.IsTruncated {
+			q.Add(queryStrKey, pDim.ID)
+		}
+		pDim.TruncateLink = generateTruncatePath(path, pDim.ID, q)
+		dimensions = append(dimensions, pDim)
+	}
+	return dimensions
+}
+
+// getTruncationMidRange returns ints that can be used as the truncation mid range
+func getTruncationMidRange(total int) (int, int) {
+	mid := total / 2
+	midFloor := mid - 2
+	midCeiling := midFloor + 3
+	return midFloor, midCeiling
+}
```

Comment:
Ideally these should be clamped to never go below zero because the values are subsequently used as index limits into an array, for which negative values don't hold. It also tightly couples the range of values emitted to the implementation of `isTruncationIndex()` which is being trusted to handle negative values.

Reply:
Indeed. This method is solely intended to ensure consistency in approach and is heavily dependent on where it is instantiated. I'll add a guard clause so at the very least the default `0` is returned should the values potentially be negative.

Reply:
Is `0, 0` what you want in this case? Or would clamping midFloor offer more reasonable behaviour? It can never be less than -2, so midCeiling is always positive.

```
midFloor := mid - 2
midCeiling := midFloor + 3
if (midFloor < 0) 
  midFloor = 0
}
return midFloor, midCeiling
```